proto/api.proto

```diff
@@ -376,3 +322,274 @@ message AdminMergeRequest {
+  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional int32 max_ranges = 2 [(gogoproto.nullable) = false];
+  // Ignore intents indicates whether or not intents encountered
+  // while looking up the range info should be resolved. This should
+  // be false in general, except for the case where the lookup is
+  // already in service of pushing intents on meta records. Attempting
+  // to resolve intents in this case would lead to infinite recursion.
+  optional bool ignore_intents = 3 [(gogoproto.nullable) = false];
+}
+
+// An RangeLookupResponse is the return value from the
+// RangeLookup() method. It returns metadata for the range
+// containing the requested key, optionally returning the metadata for
+// additional consecutive ranges beyond the requested range to pre-fill
+// the range descriptor cache.
+message RangeLookupResponse {
+  optional ResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  repeated RangeDescriptor ranges = 2 [(gogoproto.nullable) = false];
+}
+
+// An HeartbeatTxnRequest is arguments to the
+// HeartbeatTxn() method. It's sent by transaction
+// coordinators to let the system know that the transaction is still
+// ongoing. Note that this heartbeat message is different from the
+// heartbeat message in the gossip protocol.
+message HeartbeatTxnRequest {
+  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
+
+// An HeartbeatTxnResponse is the return value from the
+// HeartbeatTxn() method. It returns the transaction info in
+// the response header. The returned transaction lets the coordinator
+// know the disposition of the transaction (i.e. aborted, committed or
+// pending).
+message HeartbeatTxnResponse {
+  optional ResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
+
+// An GCRequest is arguments to the GC() method. It's
+// sent by range leaders after scanning range data to find expired
+// MVCC values.
+message GCRequest {
+  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional GCMetadata gc_meta = 2 [(gogoproto.nullable) = false, (gogoproto.customname) = "GCMeta"];
+
+  message GCKey {
+    optional bytes key = 1 [(gogoproto.casttype) = "Key"];
+    optional Timestamp timestamp = 2 [(gogoproto.nullable) = false];
+  }
+  repeated GCKey keys = 3 [(gogoproto.nullable) = false];
+}
+
+// An GCResponse is the return value from the GC()
+// method.
+message GCResponse {
+  optional ResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
+
+// TxnPushType determines what action to take when pushing a
+// transaction.
+enum PushTxnType {
+  option (gogoproto.goproto_enum_prefix) = false;
+  // Push the timestamp forward if possible to accommodate a concurrent reader.
+  PUSH_TIMESTAMP = 0;
+  // Abort the transaction if possible to accommodate a concurrent writer.
```

Comment:
OT: when is it not possible to abort the transaction? what happens if it is not possible?